src/BlockTypes/LegacyTemplate.php

```diff
@@ -180,4 +195,43 @@ protected function render_archive_product() {
 		wp_reset_postdata();
 		return ob_get_clean();
 	}
+
+	/**
+	 * Get HTML markup with the right classes by attributes.
+	 * This function appends the classname at the first element that have the class attribute.
+	 * Based on the experience, all the wrapper elements have a class attribute.
+	 *
+	 * @param string $content Block content.
+	 * @param array  $block Parsed block data.
+	 * @return string Rendered block type output.
+	 */
+	public function get_markup_with_classes_by_attributes( string $content, array $block ) {
+		if ( ! $this->is_legacy_template( $block ) ) {
+			return $content;
+		}
+
+		$pattern               = '/(?<=class=\")[^"]+(?=\")/';
```

Comment:
Looking at the `BlockTypesController::add_data_attributes`, I notice the ways we target the element there and in this method are different:

- In `BlockTypesController::add_data_attributes`, we target only the openning `<div>`. This may not work for `LegacyTemplate` with themes that use `<section>` as the wrapper tag.
- Here, we target the first element (any element, not limited to `<div>`) that contain the `class` attribute.

IMO, we need to use the same mechanism to target wrapper elements in both places. I'm tagging @mikejolley here for more insights because he originally created `BlockTypesController::add_data_attributes`.

Reply:
Ok, it makes sense!
I agree with you. This regex is not great :D
Let's wait for @mikejolley feedback!

Reply:
In the case of `add_data_attributes` we know we're dealing with a `div` element because thats how the block markup is saved to the post. If the opening element is unknown, your regex is probably better, but I wonder if you can still target only the opening element rather than the first class found which could be deeper? I am not going to pretend to be regex expert, because I am not :D

Reply:
> but I wonder if you can still target only the opening element rather than the first class found which could be deeper? I am not going to pretend to be regex expert, because I am not :D
It can be a good idea, but I'm pretty sure that the regex will be hard to write and read. I'm not a fan of complex regex, but if you think that it makes sense to work on it, I can do it!

Reply:
It might not be desirable to add that class deeper in case it breaks other parts of the layout 👍🏻 

Are there any alternatives considered for this additional class inserted by regex? Would it be a disaster to wrap in another div with the class, or does that not work here?

Reply:
> Are there any alternatives considered for this additional class inserted by regex? Would it be a disaster to wrap in another div with the class, or does that not work here?

It works without problem, but I don't know if we can wrap a block in another div. If we can follow this path, I think that it is the best solution.

![image](https://user-images.githubusercontent.com/4463174/147772127-e8d72b8a-5533-449a-8fb5-8b4359f530bc.png)

Reply:
> works without problem, but I don't know if we can wrap a block in another div. If we can follow this path, I think that it is the best solution.

I think wrapping inside another element can make it harder (and confusing) for themes developers to style the page because the opening element now is not the first element in the `wrapper-start.php` but the alignment `div`.

Reply:
With [8723818](https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/5433/commits/872381820a26354c3117323d27b551ac425e3d74), I implemented a series of regexes.

First of all, I pick the first tag HTML, after that:
- if the first tag already has the class attribute, I just add the new class.
- if the first tag doesn't have the class attribute, I add the class attribute + the new class

Let me know what do you think :D